confmap/confmap.go

```diff
@@ -143,31 +147,40 @@ func (l *Conf) ToStringMap() map[string]any {
 // uniqueness of component IDs (see mapKeyStringToMapKeyTextUnmarshalerHookFunc).
 // Decodes time.Duration from strings. Allows custom unmarshaling for structs implementing
 // encoding.TextUnmarshaler. Allows custom unmarshaling for structs implementing confmap.Unmarshaler.
-func decodeConfig(m *Conf, result any, errorUnused bool) error {
+func decodeConfig(m *Conf, result any, errorUnused bool, topLevelUnmarshaling bool) error {
+	hookFuncs := []mapstructure.DecodeHookFunc{
+		expandNilStructPointersHookFunc(),
+		mapstructure.StringToSliceHookFunc(","),
+		mapKeyStringToMapKeyTextUnmarshalerHookFunc(),
+		mapstructure.StringToTimeDurationHookFunc(),
+		mapstructure.TextUnmarshallerHookFunc(),
+		unmarshalerHookFunc(result, topLevelUnmarshaling),
+		// after the main unmarshaler hook is called,
+		// we unmarshal the embedded structs if present to merge with the result:
+		unmarshalerEmbeddedStructsHookFunc(),
+		zeroSliceHookFunc(),
+	}
```

Comment:
This is also something we can do on a separate PR!

Reply:
right, I'll revert that change. It's not helping the diff.